server/src/main/java/io/druid/server/QueryResource.java

```diff
@@ -220,8 +222,17 @@ public void write(OutputStream outputStream) throws IOException, WebApplicationE
                 },
                 contentType
             )
-            .header("X-Druid-Query-Id", queryId)
-            .header("X-Druid-Response-Context", jsonMapper.writeValueAsString(responseContext))
+            .header("X-Druid-Query-Id", queryId);
+
+        //Limit the response-context header, see https://github.com/druid-io/druid/issues/2331
+        String responseCtxString = jsonMapper.writeValueAsString(responseContext);
```

Comment:
@drcrallen @himanshug the most correct way would be to enable the `JsonGenerator .Feature.ESCAPE_NON_ASCII` feature in jackson to ensure we correctly escape all non-ascii characters, otherwise the resulting context may not be valid JSON.